### PR TITLE
Ft tictoc

### PR DESCRIPTION
**Pull request by j042**
adds Timer for easy time measurements
cherry-picked @clemens-fricke's docstring clean up

# Example use
```python
import gustaf as gus

# counting starts now!
tic = gus.utils.Tic(
    title="example Tic.toc() use",
    log_level="info" # debug is default
)

add = 1 + 2 + 3
tic.toc("addition")

mul = 1 * 2 * 3
tic.toc("multiplication")

div = 1 / 2 / 3
tic.toc("division")


# log and print summary
tic.summary(print_=True)
```
this will output
```
+++ example Tic.toc() use - time logs +++
 names          |         diff |  cummulative
 addition       | 0.0000010000 | 0.0000010000
 multiplication | 0.0000132080 | 0.0000142080
 division       | 0.0000032500 | 0.0000174580
```

### CODE REVIEW

File: gustaf/utils/__init__.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 from gustaf.utils import arr, connec, log, tictoc
```
Comment (clemens-fricke) [inline gustaf/utils/__init__.py:1]: do you need to keep the direct tictoc import in this line? if you add a renamed one in the next line?

Is this for historical reasons? should we add a deprecation warning?

File: gustaf/utils/tictoc.py [hunk 2]
```diff
@@ -2,3 +2,116 @@
...
+
+    def toc(self, name=None, log=False):
+        """
+        Adds now to the measurements.
+
+        Parameters
+        ----------
+        name: str
+          name of this specific measurement.
+          By default it assign a number to this lap.
```
Comment (clemens-fricke) [inline gustaf/utils/tictoc.py:51]:
```suggestion
          Name of this specific measurement lap.
          By default, it assigns a number to this lap.
```

File: gustaf/utils/tictoc.py [hunk 3]
```diff
@@ -2,3 +2,116 @@
...
+        """
+        # measure now
+        self._laps.append(now())
+
+        self._names.append(
+            f"lap-{len(self._laps) - 2}" if name is None else str(name)
+        )
+
+        if log:
+            self._logger("time lapsed: ", self._laps[-2] - self._laps[-1])
```
Comment (clemens-fricke) [inline gustaf/utils/tictoc.py:67]:
```suggestion
            self._logger(f"{self._title} - Lap {name}:  {self._laps[-2] - self._laps[-1]}")
```

File: gustaf/utils/tictoc.py [hunk 4]
```diff
@@ -2,3 +2,116 @@
...
+    now = now
+
+    def __init__(self, title="untitled", log_level="debug"):
+        """
+        Configures the timer and logger. Marks the starting point.
+
+        Parameters
+        ----------
+        title: str
+          title for this measurement.
```
Comment (clemens-fricke) [inline gustaf/utils/tictoc.py:27]:
```suggestion
          Title for this measurement. Defaults to "untitled".
```

File: gustaf/utils/tictoc.py [hunk 5]
```diff
@@ -2,3 +2,116 @@
...
+          lapsed time.
+        """
+        message = [f"\n+++ {self._title} - time logs +++\n"]
+
+        name_width = int(max([len(n) for n in self._names]))
+
+        # extract info
+        start = self._laps[0]
+        cummulative_raw = [lap - start for lap in self._laps[1:]]
+        cummulative = [f"{c:.10f}" for c in cummulative_raw]
```
Comment (clemens-fricke) [inline gustaf/utils/tictoc.py:91]:
```suggestion
        cummulative = [f"{lap-start:.10f}" for lap in self._laps[1:]]
```
Why not keep it in one line?

**Review by clemens-fricke — CHANGES_REQUESTED** [review]

File: gustaf/utils/tictoc.py [hunk 6]
```diff
@@ -2,3 +2,116 @@
...
+    def __init__(self, title="untitled", log_level="debug"):
+        """
+        Configures the timer and logger. Marks the starting point.
+
+        Parameters
+        ----------
+        title: str
+          title for this measurement.
+        log_level: str
+          Valid options are {"info", "debug", "warning"}
```
Comment (clemens-fricke) [inline gustaf/utils/tictoc.py:29]:
```suggestion
          Valid options are {"info", "debug", "warning"}. Defaults to "debug".
```

File: gustaf/utils/tictoc.py [hunk 7]
```diff
@@ -2,3 +2,116 @@
...
+        """
+        Adds now to the measurements.
+
+        Parameters
+        ----------
+        name: str
+          name of this specific measurement.
+          By default it assign a number to this lap.
+        log: bool
+          If True, will log lapsed time.
```
Comment (clemens-fricke) [inline gustaf/utils/tictoc.py:53]:
```suggestion
          If True, will log lapsed time. Defaults to None.
```

File: gustaf/utils/tictoc.py [hunk 8]
```diff
@@ -2,3 +2,116 @@
...
+        Parameters
+        ----------
+        print_: bool
+
+
+        Returns
+        -------
+        records: tuple
+          Records starting from starting point. tuple of names and cummulative
+          lapsed time.
```
Comment (clemens-fricke) [inline gustaf/utils/tictoc.py:82]:
```suggestion
        Prints measurement summary to the log.

        Parameters
        ----------
        print_: Print the summary also to stdout. Defaults to False.


        Returns
        -------
        records: tuple
          Lap timings in a tuple consisting of a list of the lap names and
           times for each lap from the timer start (cumulative time).
```

File: gustaf/utils/tictoc.py [hunk 9]
```diff
@@ -2,3 +2,116 @@
...
+        self._laps.append(now())
+
+        self._names.append(
+            f"lap-{len(self._laps) - 2}" if name is None else str(name)
+        )
+
+        if log:
+            self._logger("time lapsed: ", self._laps[-2] - self._laps[-1])
+
+    def summary(self, print_=False):
```
Comment (clemens-fricke) [inline gustaf/utils/tictoc.py:69]: Maybe also add an option of getting the data without having the log file filled? Maybe, a Getter function for the lap times?

def timeings(self)->List[Tuple[str, time]] or something like this.

File: gustaf/utils/__init__.py (hunk 1 above)
Comment (j042) [reply gustaf/utils/__init__.py:1]: yes

**Pull request merged**